polyval/src/backend/clmul.rs

```diff
@@ -155,10 +151,143 @@ impl Reset for Polyval {
 }
 
 #[cfg(feature = "zeroize")]
-impl Drop for Polyval {
+impl<const N: usize> Drop for Polyval<N> {
     fn drop(&mut self) {
         use zeroize::Zeroize;
         self.h.zeroize();
         self.y.zeroize();
     }
 }
+
+/// # Safety
+///
+/// The SSE2 and pclmulqdq target features must be enabled.
+#[inline]
+#[target_feature(enable = "sse2,pclmulqdq")]
+#[allow(unused_unsafe)]
+#[allow(clippy::undocumented_unsafe_blocks, reason = "Too many unsafe blocks.")]
+unsafe fn polymul(x: __m128i, y: __m128i) -> __m128i {
+    let (h, m, l) = unsafe { karatsuba1(x, y) };
+    let (h, l) = unsafe { karatsuba2(h, m, l) };
+    unsafe {
+        mont_reduce(h, l) // d
+    }
+}
+
+/// Karatsuba decomposition for `x*y`.
+#[inline]
+#[target_feature(enable = "sse2,pclmulqdq")]
+#[allow(unused_unsafe)]
+#[allow(clippy::undocumented_unsafe_blocks, reason = "Too many unsafe blocks.")]
+unsafe fn karatsuba1(x: __m128i, y: __m128i) -> (__m128i, __m128i, __m128i) {
+    // First Karatsuba step: decompose x and y.
+    //
+    // (x1*y0 + x0*y1) = (x1+x0) * (y1+x0) + (x1*y1) + (x0*y0)
+    //        M                                 H         L
+    //
+    // m = x.hi^x.lo * y.hi^y.lo
+    let m = unsafe {
+        pmull(
+            _mm_xor_si128(x, _mm_shuffle_epi32(x, 0xee)),
+            _mm_xor_si128(y, _mm_shuffle_epi32(y, 0xee)),
+        )
+    };
+    let h = unsafe { pmull2(y, x) }; // h = x.hi * y.hi
+    let l = unsafe { pmull(y, x) }; // l = x.lo * y.lo
+    (h, m, l)
+}
+
+/// Karatsuba combine.
+#[inline]
+#[target_feature(enable = "sse2,pclmulqdq")]
+#[allow(unused_unsafe)]
+#[allow(clippy::undocumented_unsafe_blocks, reason = "Too many unsafe blocks.")]
```

Comment:
Why are there `unsafe` blocks inside an `unsafe fn`?

Reply:
Rust somewhat recently made it a lint error to not use unsafe blocks inside unsafe functions.

Reply:
https://rust-lang.github.io/rfcs/2585-unsafe-block-in-unsafe-fn.html

Reply:
In other crates we just disabled this lint for relevant modules, e.g. like [here](https://github.com/RustCrypto/block-ciphers/blob/master/aes/src/ni/encdec.rs#L1).